telegram/ext/builders.py

```diff
@@ -0,0 +1,1338 @@
+        return self
+
+    @check_if_already_set
+    def _set_token(self: BuilderType, token: str) -> BuilderType:
+        if self._bot_was_set:
+            raise self._exception_builder('token', 'bot instance')
+        if self._dispatcher_check:
+            raise self._exception_builder('token', 'Dispatcher instance')
+        self._token = token
+        self._token_was_set = True
+        return self
+
+    @check_if_already_set
+    def _set_base_url(self: BuilderType, base_url: str) -> BuilderType:
+        if self._bot_was_set:
+            raise self._exception_builder('base_url', 'bot instance')
+        if self._dispatcher_check:
+            raise self._exception_builder('base_url', 'Dispatcher instance')
+        self._base_url = base_url
+        self._base_url_was_set = True
+        return self
+
+    @check_if_already_set
+    def _set_base_file_url(self: BuilderType, base_file_url: str) -> BuilderType:
+        if self._bot_was_set:
+            raise self._exception_builder('base_file_url', 'bot instance')
+        if self._dispatcher_check:
+            raise self._exception_builder('base_file_url', 'Dispatcher instance')
+        self._base_file_url = base_file_url
+        self._base_file_url_was_set = True
+        return self
+
+    @check_if_already_set
+    def _set_request_kwargs(self: BuilderType, request_kwargs: Dict[str, Any]) -> BuilderType:
+        if self._request_was_set:
+            raise self._exception_builder('request_kwargs', 'Request instance')
+        if self._bot_was_set:
+            raise self._exception_builder('request_kwargs', 'bot instance')
+        if self._dispatcher_check:
+            raise self._exception_builder('request_kwargs', 'Dispatcher instance')
+        self._request_kwargs = request_kwargs
+        self._request_kwargs_was_set = True
+        return self
+
+    @check_if_already_set
+    def _set_request(self: BuilderType, request: Request) -> BuilderType:
+        if self._request_kwargs_was_set:
+            raise self._exception_builder('request', 'request_kwargs')
+        if self._bot_was_set:
+            raise self._exception_builder('request', 'bot instance')
+        if self._dispatcher_check:
+            raise self._exception_builder('request', 'Dispatcher instance')
+        self._request = request
+        self._request_was_set = True
+        return self
+
+    @check_if_already_set
+    def _set_private_key(self: BuilderType, private_key: bytes) -> BuilderType:
+        if self._bot_was_set:
+            raise self._exception_builder('private_key', 'bot instance')
+        if self._dispatcher_check:
+            raise self._exception_builder('private_key', 'Dispatcher instance')
+        self._private_key = private_key
+        self._private_key_was_set = True
+        return self
+
+    @check_if_already_set
+    def _set_private_key_password(self: BuilderType, private_key_password: bytes) -> BuilderType:
+        if self._bot_was_set:
```

Comment:
Should we assert that `self._request` wasn't set as well?

Reply:
https://github.com/python-telegram-bot/python-telegram-bot/pull/2646#discussion_r706642955